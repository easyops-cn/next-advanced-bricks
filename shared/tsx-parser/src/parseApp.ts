import type { ParsedApp, SourceFile } from "./modules/interfaces.js";
import { parseFile } from "./modules/parseFile.js";

export function parseApp(files: SourceFile[]) {
  const app: ParsedApp = {
    appType: "app",
    modules: new Map(),
    files,
    errors: [],
  };

  parseFile("/index.tsx", app);

  return app;
}
