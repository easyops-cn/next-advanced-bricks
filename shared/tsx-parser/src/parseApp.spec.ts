import { readFile } from "node:fs/promises";
import path from "node:path";
import { parseApp } from "./parseApp.js";

describe("parseApp", () => {
  test("should work", async () => {
    const indexContent = await readFile(
      path.join(__dirname, "./__fixtures__/app/index.txt"),
      "utf-8"
    );
    const testContent = await readFile(
      path.join(__dirname, "./__fixtures__/app/Pages/Test.txt"),
      "utf-8"
    );

    const files = [
      {
        filePath: "/index.tsx",
        content: indexContent,
      },
      {
        filePath: "/Pages/Test.tsx",
        content: testContent,
      },
    ];

    const app = parseApp(files);
    const simplifiedApp = {
      ...app,
      files: undefined,
      entry: app.entry
        ? {
            ...app.entry,
            source: undefined,
          }
        : app.entry,
      modules: new Map(
        Array.from(app.modules.entries()).map(([k, v]) => [
          k,
          v
            ? {
                ...v,
                source: undefined,
              }
            : v,
        ])
      ),
    };
    expect(simplifiedApp).toMatchSnapshot();
  });
});
