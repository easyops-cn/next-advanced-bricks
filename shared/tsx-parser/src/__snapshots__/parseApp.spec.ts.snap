// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`parseApp should work 1`] = `
{
  "appType": "app",
  "entry": {
    "contracts": Set {},
    "defaultExport": null,
    "errors": [],
    "filePath": "index.tsx",
    "internals": [
      {
        "function": {
          "name": "sayHello",
          "source": "function sayHello(name: string) {
  return \`Hello, \${name}!\`;
}",
          "typescript": true,
        },
        "type": "function",
      },
    ],
    "moduleType": "entry",
    "namedExports": Map {},
    "render": {
      "children": [
        {
          "children": [
            {
              "children": undefined,
              "events": undefined,
              "name": "Route",
              "properties": {
                "component": {
                  "name": "Layout1",
                  "reference": {
                    "importSource": "./Pages/Test",
                    "type": "imported",
                  },
                },
                "path": "/",
              },
              "ref": undefined,
              "reference": {
                "importSource": "next-tsx",
                "name": "Route",
                "type": "imported",
              },
            },
            {
              "children": undefined,
              "events": undefined,
              "name": "Route",
              "properties": {
                "component": {
                  "name": "Page403",
                  "reference": {
                    "importSource": "./Pages/Test",
                    "name": "Page503",
                    "type": "imported",
                  },
                },
                "path": "/404",
              },
              "ref": undefined,
              "reference": {
                "importSource": "next-tsx",
                "name": "Route",
                "type": "imported",
              },
            },
          ],
          "events": undefined,
          "name": "Routes",
          "properties": {},
          "ref": undefined,
          "reference": {
            "importSource": "next-tsx",
            "name": "Routes",
            "type": "imported",
          },
        },
      ],
      "type": "render",
    },
    "source": undefined,
    "usedHelpers": Set {},
  },
  "errors": [],
  "files": undefined,
  "modules": Map {
    "index.tsx" => {
      "contracts": Set {},
      "defaultExport": null,
      "errors": [],
      "filePath": "index.tsx",
      "internals": [
        {
          "function": {
            "name": "sayHello",
            "source": "function sayHello(name: string) {
  return \`Hello, \${name}!\`;
}",
            "typescript": true,
          },
          "type": "function",
        },
      ],
      "moduleType": "entry",
      "namedExports": Map {},
      "render": {
        "children": [
          {
            "children": [
              {
                "children": undefined,
                "events": undefined,
                "name": "Route",
                "properties": {
                  "component": {
                    "name": "Layout1",
                    "reference": {
                      "importSource": "./Pages/Test",
                      "type": "imported",
                    },
                  },
                  "path": "/",
                },
                "ref": undefined,
                "reference": {
                  "importSource": "next-tsx",
                  "name": "Route",
                  "type": "imported",
                },
              },
              {
                "children": undefined,
                "events": undefined,
                "name": "Route",
                "properties": {
                  "component": {
                    "name": "Page403",
                    "reference": {
                      "importSource": "./Pages/Test",
                      "name": "Page503",
                      "type": "imported",
                    },
                  },
                  "path": "/404",
                },
                "ref": undefined,
                "reference": {
                  "importSource": "next-tsx",
                  "name": "Route",
                  "type": "imported",
                },
              },
            ],
            "events": undefined,
            "name": "Routes",
            "properties": {},
            "ref": undefined,
            "reference": {
              "importSource": "next-tsx",
              "name": "Routes",
              "type": "imported",
            },
          },
        ],
        "type": "render",
      },
      "source": undefined,
      "usedHelpers": Set {},
    },
    "Pages/Test.tsx" => {
      "contracts": Set {},
      "defaultExport": {
        "component": {
          "bindingMap": Map {},
          "children": [
            {
              "children": undefined,
              "events": undefined,
              "name": "h1",
              "properties": {
                "textContent": "Home",
              },
              "ref": undefined,
              "reference": null,
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": undefined,
                          "events": undefined,
                          "name": "Link",
                          "properties": {
                            "textContent": "Home",
                            "to": "/",
                          },
                          "ref": undefined,
                          "reference": null,
                        },
                      ],
                      "events": undefined,
                      "name": "li",
                      "properties": {},
                      "ref": undefined,
                      "reference": null,
                    },
                    {
                      "children": [
                        {
                          "children": undefined,
                          "events": undefined,
                          "name": "Link",
                          "properties": {
                            "textContent": "About",
                            "to": "/about",
                          },
                          "ref": undefined,
                          "reference": null,
                        },
                      ],
                      "events": undefined,
                      "name": "li",
                      "properties": {},
                      "ref": undefined,
                      "reference": null,
                    },
                  ],
                  "events": undefined,
                  "name": "ul",
                  "properties": {},
                  "ref": undefined,
                  "reference": null,
                },
              ],
              "events": undefined,
              "name": "Nav",
              "properties": {},
              "ref": undefined,
              "reference": {
                "name": "Nav",
                "type": "local",
              },
            },
            {
              "children": [
                {
                  "children": undefined,
                  "events": undefined,
                  "name": "h2",
                  "properties": {
                    "slot": "header",
                    "textContent": "Welcome",
                  },
                  "ref": undefined,
                  "reference": null,
                },
                {
                  "children": [
                    {
                      "children": undefined,
                      "events": undefined,
                      "name": "Route",
                      "properties": {
                        "path": "/",
                      },
                      "ref": undefined,
                      "reference": null,
                    },
                    {
                      "children": undefined,
                      "events": undefined,
                      "name": "Route",
                      "properties": {
                        "path": "/about",
                      },
                      "ref": undefined,
                      "reference": null,
                    },
                  ],
                  "events": undefined,
                  "name": "Routes",
                  "properties": {
                    "slot": "body",
                  },
                  "ref": undefined,
                  "reference": null,
                },
              ],
              "events": undefined,
              "name": "Card",
              "properties": {},
              "ref": undefined,
              "reference": null,
            },
          ],
          "id": Node {
            "end": 67,
            "loc": SourceLocation {
              "end": Position {
                "column": 30,
                "index": 67,
                "line": 4,
              },
              "filename": undefined,
              "identifierName": "Layout",
              "start": Position {
                "column": 24,
                "index": 61,
                "line": 4,
              },
            },
            "name": "Layout",
            "start": 61,
            "type": "Identifier",
          },
          "type": "page",
        },
        "type": "page",
      },
      "errors": [
        {
          "message": "The component "Home" is not defined",
          "node": Node {
            "end": 426,
            "loc": SourceLocation {
              "end": Position {
                "column": 41,
                "index": 426,
                "line": 21,
              },
              "filename": undefined,
              "identifierName": "Home",
              "start": Position {
                "column": 37,
                "index": 422,
                "line": 21,
              },
            },
            "name": "Home",
            "start": 422,
            "type": "Identifier",
          },
          "severity": "error",
        },
        {
          "message": "The component "About" is not defined",
          "node": Node {
            "end": 478,
            "loc": SourceLocation {
              "end": Position {
                "column": 47,
                "index": 478,
                "line": 22,
              },
              "filename": undefined,
              "identifierName": "About",
              "start": Position {
                "column": 42,
                "index": 473,
                "line": 22,
              },
            },
            "name": "About",
            "start": 473,
            "type": "Identifier",
          },
          "severity": "error",
        },
      ],
      "filePath": "Pages/Test.tsx",
      "internals": [
        {
          "component": {
            "bindingMap": Map {},
            "children": [
              {
                "children": [
                  {
                    "children": undefined,
                    "events": undefined,
                    "name": "slot",
                    "properties": {},
                    "ref": undefined,
                    "reference": null,
                  },
                ],
                "events": undefined,
                "name": "nav",
                "properties": {},
                "ref": undefined,
                "reference": null,
              },
            ],
            "id": Node {
              "end": 650,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "index": 650,
                  "line": 35,
                },
                "filename": undefined,
                "identifierName": "Nav",
                "start": Position {
                  "column": 9,
                  "index": 647,
                  "line": 35,
                },
              },
              "name": "Nav",
              "start": 647,
              "type": "Identifier",
            },
            "type": "template",
          },
          "type": "template",
        },
      ],
      "moduleType": "page",
      "namedExports": Map {
        "Page404" => {
          "component": {
            "bindingMap": Map {},
            "children": [
              {
                "children": undefined,
                "events": undefined,
                "name": "h1",
                "properties": {
                  "textContent": "404 Not Found",
                },
                "ref": undefined,
                "reference": null,
              },
            ],
            "id": Node {
              "end": 575,
              "loc": SourceLocation {
                "end": Position {
                  "column": 23,
                  "index": 575,
                  "line": 30,
                },
                "filename": undefined,
                "identifierName": "Page404",
                "start": Position {
                  "column": 16,
                  "index": 568,
                  "line": 30,
                },
              },
              "name": "Page404",
              "start": 568,
              "type": "Identifier",
            },
            "type": "template",
          },
          "type": "template",
        },
      },
      "source": undefined,
      "usedHelpers": Set {},
    },
  },
}
`;
